### PR TITLE
Add csp to html 404 page

### PR DESCRIPTION
**Pull request by g-k**
refs: [get FxA passing the webappsec baseline](https://github.com/mozilla-services/foxsec/issues/177) (should be the last application level content-server change)

Note: the 404 page doesn't include any JS and could be served with a much more locked down header, but I figured it'd to keep things simpler to use the default CSP config.

Functional Tests:

- [x] 404 page includes CSP header for non-html path

```
» curl -IX GET http://localhost:3030/sitemap.xml
HTTP/1.1 404 Not Found
X-XSS-Protection: 1; mode=block
Strict-Transport-Security: max-age=15552000; includeSubDomains
X-Content-Type-Options: nosniff
Content-Type: text/html; charset=utf-8
Content-Security-Policy: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114; d
efault-src 'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-sr
c 'none'; report-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
X-Content-Security-Policy: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114;
 default-src 'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-
src 'none'; report-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
X-WebKit-CSP: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114; default-src
'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-src 'none'; r
eport-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
Content-Length: 24
Date: Thu, 16 Feb 2017 21:01:26 GMT
Connection: keep-alive
```

- [x] non-html 404 page does not include CSP header

```
» curl -H 'Accept: application/json' -IX GET http://localhost:3030/sitemap.xml
HTTP/1.1 404 Not Found
X-XSS-Protection: 1; mode=block
Strict-Transport-Security: max-age=15552000; includeSubDomains
X-Content-Type-Options: nosniff
Content-Type: application/json; charset=utf-8
Content-Length: 21
ETag: W/"15-Lvrg4Gz9w+rMdDVG/1lddA"
Date: Thu, 16 Feb 2017 21:02:40 GMT
Connection: keep-alive
```

- [x] other page returns CSP as usual

```
» curl -IX GET http://localhost:3030/
HTTP/1.1 200 OK
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block
Strict-Transport-Security: max-age=15552000; includeSubDomains
X-Content-Type-Options: nosniff
Content-Security-Policy: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114; d
efault-src 'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-sr
c 'none'; report-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
X-Content-Security-Policy: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114;
 default-src 'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-
src 'none'; report-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
X-WebKit-CSP: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114; default-src
'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-src 'none'; r
eport-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
X-Robots-Tag: noindex,nofollow
Content-Type: text/html; charset=utf-8
Content-Length: 3472
ETag: W/"d90-Dh4DnXs0aQ8hhLURkG5X/Q"
Date: Thu, 16 Feb 2017 21:03:56 GMT
Connection: keep-alive
```

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: d2f4cefe681c580a028724395a7294a12f6214a2
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 32a5f5ac9a71fcc06f560fe4a07a8767f0ea1d05
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by shane-tomlinson** [comment]
Thanks for the PR @g-k. The approach looks reasonable, though the 404 page no longer renders.

STR

1) visit https://127.0.0.1:3030/404

### Existing

screen shot 2017-02-21 at 13 58 37

### With this PR

screen shot 2017-02-21 at 13 59 06

File: server/lib/404.js [hunk 1]
```diff
@@ -8,7 +8,8 @@ module.exports = function (req, res, next) {
   res.status(404);
 
   if (req.accepts('html')) {
-    return res.render('404');
+    res.setHeader('content-type', 'text/html; charset=utf-8');
```
Comment (shane-tomlinson) [inline server/lib/404.js]: I think the problem is the CSP headers are set after the page is rendered, which IIUC essentially resets the render. So, we might have to set the `content-type` header, then call the CSP middleware, then do the render.

**Comment by g-k** [comment]
Ah, I just needed to [remove the render callback to next](https://github.com/mozilla/fxa-content-server/pull/4729/files#diff-2f86b9285db11c4335722905c2554e9eR12
) since it's for template engine post processing and needs to return html (which the middleware next  never does).

https://expressjs.com/en/4x/api.html#res.render

```
» curl -vX GET http://localhost:3030/notfound
*   Trying ::1...
* connect to ::1 port 3030 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3030 (#0)
> GET /notfound HTTP/1.1
> Host: localhost:3030
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< X-Frame-Options: DENY
< X-XSS-Protection: 1; mode=block
< Strict-Transport-Security: max-age=15552000; includeSubDomains
< X-Content-Type-Options: nosniff
< Content-Security-Policy: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114; default-src 'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-src 'none'; report-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
< X-Content-Security-Policy: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114; default-src 'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-src 'none'; report-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
< X-WebKit-CSP: connect-src 'self' http://127.0.0.1:9000 http://127.0.0.1:9010 http://127.0.0.1:1111 http://127.0.0.1:1114; default-src 'self'; font-src 'self'; img-src 'self' data: https://secure.gravatar.com http://127.0.0.1:1112; media-src blob:; object-src 'none'; report-uri /_/csp-violation; script-src 'self'; style-src 'self' 'sha256-9n6ek6ecEYlqel7uDyKLy6fdGNo3vw/uScXSq9ooQlk='
< X-Robots-Tag: noindex,nofollow
< Content-Type: text/html; charset=utf-8
< Content-Length: 1649
< ETag: W/"671-hjHh+LVo9z7gqDw1fvusdw"
< Date: Tue, 21 Feb 2017 19:34:33 GMT
< Connection: keep-alive
< 
{ [1649 bytes data]

100  1649  100  1649    0     0  70851      0 --:--:-- --:--:-- --:--:-- 78523
* Connection #0 to host localhost left intact
<!doctype html>
<!--[if lt IE 10]>       <html class="lt-ie10" dir="ltr" lang="en"> <![endif]-->
<!--[if gte IE 10]><!--> <html dir="ltr" lang="en"> <!--<![endif]-->
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title>Firefox Accounts</title>
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width">
        <meta name="referrer" content="origin">
        <meta name="robots" content="noindex,nofollow">

        <!-- build:css(.tmp) /styles/en.css -->
        <link rel="stylesheet" href="/bower_components/normalize-css/normalize.css">
        <link rel="stylesheet" href="/styles/fontello.css">
        <link rel="stylesheet" href="/styles/fonts/en.css">
        <link rel="stylesheet" href="/styles/main.css">
        <!-- endbuild -->
    </head>
    <body class="static" data-static-resource-host="">
        <div id="fox-logo"></div>
        <div id="stage">
            <div id="main-content" class="card">
              <header>
                <h1 id="fxa-404-header">Page not found</h1>
              </header>

              <section>
                The page you requested is not found. We've been notified and will fix any links that may be broken.
              </section>

              <div class="button-row">
                <a href="/signup" class="button" id="fxa-404-home">Home</a>
              </div>
            </div>
        </div>

        <a id="about-mozilla" rel="author" target="_blank" href="https://www.mozilla.org/about/?utm_source=firefox-accounts&amp;utm_medium=Referral"></a>
    </body>
</html>
```

screen shot 2017-02-21 at 2 31 29 pm

**Comment by shane-tomlinson** [comment]
@g-k thanks for the update. Unfortunately, we have another problem now - the header isn't set whenever the request contains a non-HTML extension. Here's a diff showing the problem:

```diff
diff --git a/tests/server/routes.js b/tests/server/routes.js
index 497c8243..095b9833 100644
--- a/tests/server/routes.js
+++ b/tests/server/routes.js
@@ -96,10 +96,11 @@ define([
   }
 
   if (! intern.config.fxaProduction) {
-    routes['/tests/index.html'] = { statusCode: 200 };
-    routes['/tests/index.html?coverage'] = { statusCode: 200 };
+    routes['/tests/index.html'] = { csp: false, statusCode: 200 };
+    routes['/tests/index.html?coverage'] = { csp: false, statusCode: 200 };
     routes['/boom'] = { statusCode: 500 };
     routes['/non_existent'] = { statusCode: 404 };
+    routes['/non_existent.js'] = { statusCode: 404 };
     routes['/legal/non_existent'] = { statusCode: 404 };
     routes['/en/legal/non_existent'] = { statusCode: 404 };
   }
@@ -203,16 +204,15 @@ define([
 
     // all HTML pages by default have x-frame-options: DENY
     if (headers['content-type'].indexOf('text/html') > -1) {
-      assert.equal(headers['x-frame-options'], 'DENY');
-    }
+    //  assert.equal(headers['x-frame-options'], 'DENY');
 
-    // fxa-dev boxes currently do not set CSP headers (but should - GH-2155)
-    var routePathname = url.parse(route).pathname;
-    if (csp.isCspRequired(routePathname)) {
-      assert.ok(headers.hasOwnProperty('content-security-policy-report-only'));
-      assert.ok(headers.hasOwnProperty('content-security-policy'));
+      if (routes[route].csp !== false) {
+        //assert.ok(headers.hasOwnProperty('content-security-policy-report-only'));
+        assert.ok(headers.hasOwnProperty('content-security-policy'));
+      }
     }
 
+    var routePathname = url.parse(route).pathname;
     // All non-json routes in this test should have an x-robots-tag header.
     if (routes[routePathname].headerAccept !== 'application/json') {
       assert.equal(headers['x-robots-tag'], 'noindex,nofollow');
```

I'm surprised the existing text for `report-only` has not been failing, it should be because that header is not currently set. That leads me to believe that csp.isCspRequired is always returning false in the unit tests.

I've been working on a change today that sets the headers on all HTML responses by examining the `content-type` header universally, but it causes a lot of unit test modifications. I'm going to put a bit more effort into it to see where it goes. I'll let you know if it's successful.

**Comment by g-k** [comment]
Are we overthinking this?

I think it'd be OK to return the CSP header with non-html pages or call setHeader from the 404 handler with a hardcoded CSP to avoid the second CSP middleware.

**Comment by shane-tomlinson** [comment]
@g-k - see https://github.com/mozilla/fxa-content-server/pull/4750

**Comment by g-k** [comment]
👍 nice! Closing this in favor of that branch.

**Pull request closed without merging**